### PR TITLE
Fix writef() / readf() for abstract and semi-abstract enums

### LINKED ISSUE #12344: Abstract enums emit cast-to-int error when used in formatted I/O
Bug? Abstract enums are not coerced properly when used in formatted I/O.

```chapel
proc print(x) {
  writeln(x);

  var s1 = "%s".format(x);
  writeln(s1);

  var s2 = "%t".format(x);
  writeln(s2);
}

enum E { E1 };
var e = E.E1;
print(e);

enum F { F1 = 0 };
var f = F.F1;
print(f);
```
For e, s1 and s2 do not compile --
```terminal
$CHPL_HOME/modules/standard/IO.chpl:5481: error: can't cast from abstract enum ('E') to int(64)
```
For f, it works fine and prints the following:
```terminal
F1
F1
F1
```

### PR DESCRIPTION
**Pull request by bradcray**
It seems that our code paths for writef() and readf() calls were trying to
cast between enums and other types which is no longer supported for
enums without underlying values (e.g., abstract enums).  Worse, because
the current module logic does all of its conversions at execution time,
we'd get an "illegal cast" error even if the user was doing a completely
legal/reasonable writef() because the module generates the code
necessary to handle all possible conversions that it might encounter at
execution time.

Here, I special-cased enums along such code paths by generating an
error condition for abstract enums similar to other conversion-oriented
errors that we generate, and by taking more care when
reading enums in to cast them to integers before casting to their
ultimate type.

This PR follows the current approach taken for writef() (and readf()),
which is to cast if the % format doesn't match the actual argument
type.  In issue #12423 I ask whether this behavior is what we want
or not (it seemed surprising to me) for future consideration.

Resolves #12344

### CODE REVIEW

**Comment by bradcray** [comment]
Hi @mppf — You understand the writef() code better than I do, so I'm wondering whether you could review this and whether you have a thought as to whether we could somehow statically avoid sending enums down these code paths to begin with (since it's not clear to me that they're being used... or maybe it's just not clear what it would take to use them).  Thanks!

**Comment by bradcray** [comment]
Oh, upon thinking about it overnight, maybe I understand these code paths better.  Are they triggered when, say, we use a `"%i"` format string as a means of saying "convert this argument to an integer?"  In which case, we should probably make all of these conversions return the "error tuple" for enums since they don't coerce to numeric values anymore.  This would permit the code paths to compile (no halts) while preventing the conversions that don't make sense?

**Comment by mppf** [comment]
Looks like a good start to me.

Could you:
 * return `false` as the second tuple element in the event that the enum isn't supposed to cast to/from integral types, as you proposed
 * also update _setIfPrimitive which casts int -> enum (and currently fails to compile for readf on an abstract enum). For this one you can return ERANGE if the enum is not castable to/from integer.
 * update the test / add a test to include some readf calls on abstract enums
 * update the test / add a test to cover the case in which we want an error when reading the enum

Thanks!

**Comment by bradcray** [comment]
OK, I've added the tests of read behavior as well now, so I think this is ready for review and merge if it passes.

File: modules/standard/IO.chpl [hunk 1]
```diff
@@ -5754,7 +5775,17 @@ private inline
 proc _setIfPrimitive(ref lhs:?t, rhs:?t2, argi:int):syserr where t!=bool&&_isIoPrimitiveType(t)
 {
   try {
-    lhs = rhs:t;
+    if isAbstractEnumType(t) {
+      return ERANGE;
```
Comment (mppf) [inline modules/standard/IO.chpl]: I would think you could read an abstract enum from a string? So should this check be in an `isIntType(t2)` block?

Comment (mppf) [reply modules/standard/IO.chpl]: I think we don't actually yet have a test of a totally abstract enum and reading it with %s.

Comment (bradcray) [reply modules/standard/IO.chpl]: Good catch, thanks!

**Comment by bradcray** [comment]
I've just added a commit to handle and test the case you identified if you'd be up for reviewing just that commit.  I'm kicking off testing now, but don't expect surprises.

**Pull request merged**